### DISCUSSION TITLE
CefSharp opens the Chromium browser when opening more than 1 instances of the program

### DISCUSSION THREAD

**Discussion by jkiller295** [cefsharp/CefSharp#4978] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

130.1.90

### Operating System

Windows 11

### Architecture

x86

### .Net Version

.NET Framework 4.7

### Implementation

WinForms

### Reproduction Steps

1.  Install CefSharp.Winforms 130.1.90 from Nuget
2. Add the following code to the `Program.cs` file, just above the line `Application.EnableVisualStyles();`
```csharp
CefSettings settings = new CefSettings();
settings.LogSeverity = LogSeverity.Disable;
Cef.Initialize(settings);
```
3. Build the project. Open more than 1 instance of the program

### Expected behavior

All the instances of the program should open normally.

### Actual behavior

From the 2nd instance onwards, a Chromium browser opens with an error. FYI it is happening to the CefSharp.MinimalExample.WinForms sample solution downloaded from here too.
![image](https://github.com/user-attachments/assets/605687d0-b327-42f9-b9f9-780a57c8880a)

### Regression?

It worked fine in 120.2.70 and up to 126.2.180.0. 
127.3.50 onwards have this issue

### Does this problem also occur in the CEF Sample Application

Not Tested

**Comment by amaitland** [comment]
Converted to discussion as this is not a bug, it's a change in `upstream` behavior. 

> All the instances of the program should open normally.

CEF now supports the Chromium process singleton, this brings with it changes to how CachePath/RootCachePath are accessed from multiple processes. See https://github.com/cefsharp/CefSharp/issues/4668 and https://github.com/chromiumembedded/cef/issues/3609 for more details

What you are seeing is the expected behavior based on the example you've provided. 

If `Cef.Initialize` returns false then you cannot continue.

**Reply by jkiller295** [reply]
It's just strange because all versions before 126.2.180.0 were working without any issues. I can't say it was luck because I've been using that version for years. Decided to upgrade to the latest version or anything after 126.2.180.0 and boom, problem. 

How should I run multiple instances of my application? Or how do I change it to use in-memory cache like before?

**Reply by amaitland** [reply]
> It's just strange because all versions before 126.2.180.0 were working without any issues. I can't say it was luck because I've been using that version for years. Decided to upgrade to the latest version or anything after 126.2.180.0 and boom, problem.

Not strange at all. `CEF` made major changes, `126` is the first version to switch to the `Chrome Bootstrap`.

https://github.com/cefsharp/CefSharp/issues/4835

> How should I run multiple instances of my application?

You need unique `CefSettings.RootCachePath` per application instance.

**Reply by jkiller295** [reply]
Does it have to be unique for the `RootCachePath` or just `CefCachePath` will do?

**Reply by amaitland** [reply]
You need a unique `RootCachePath` per application instance.

You can load `chrome://version` to see what that is.

**Comment by MarkHaroldson** [comment]
I can work around the issue. Is there a way to override the error browser from even displaying? It seems to appear during the call the Cef.Initialize(). Even if the program detects that Initialized failed there does not appear to be any way to prevent the browser from displaying.

**Reply by amaitland** [reply]
> Return true if the relaunch is handled or false for default relaunch behavior. Default behavior will create a new default styled Chrome window.

https://cefsharp.github.io/api/130.1.x/html/M_CefSharp_IBrowserProcessHandler_OnAlreadyRunningAppRelaunch.htm

The default `upstream` behaviour is to launch a new `Chrome styled` window. You can `return true` in `IBrowserProcessHandler.OnAlreadyRunningAppRelaunch` to override this behavior.

**Reply by MarkHaroldson** [reply]
Thanks! 

What i figured out is that IBrowserProcessHandler.OnAlreadyRunningAppRelaunch is called in the context of the process that called Cef.Initialize() with the RootCachePath first and not the subsequent calls to Cef.Initialize().

**Reply by amaitland** [reply]
> What i figured out is that IBrowserProcessHandler.OnAlreadyRunningAppRelaunch is called in the context of the process that called Cef.Initialize() with the RootCachePath first and not the subsequent calls to Cef.Initialize().

Correct. This allows you to implement the same behavior as `Chromium`.

Subsequent process launches communicates with first process saying open a new window, subsequent process closes. This `New Window` is what `CEF` launches by default matching the `Chromium` behavior.
